pennylane/estimator/resource_tracking.py

```diff
@@ -0,0 +1,410 @@
+            when they are not specified on the operator.
+
+    Returns:
+        Resources: the quantum resources required to execute the circuit
+
+    Raises:
+        TypeError: could not obtain resources for obj of type :code:`type(obj)`
+
+    **Example**
+
+    We can track the resources of a quantum workflow by passing the quantum function defining our
+    workflow directly into this function.
+
+    .. code-block:: python
+
+        import pennylane.labs.resource_estimation as plre
+
+        def my_circuit():
+            for w in range(2):
+                plre.ResourceHadamard(wires=w)
+
+            plre.ResourceCNOT(wires=[0,1])
+            plre.ResourceRX(wires=0)
+            plre.ResourceRY(wires=1)
+
+            plre.ResourceQFT(num_wires=3, wires=[0, 1, 2])
+            return
+
+    Note that we are passing a python function NOT a :class:`~.QNode`. The resources for this
+    workflow are then obtained by:
+
+    >>> config = plre.ResourceConfig()
+    >>> config.set_single_qubit_rot_precision(1e-4)
+    >>> res = plre.estimate(
+    ...     my_circuit,
+    ...     gate_set = plre.DefaultGateSet,
+    ...     config=config,
+    ... )()
+    ...
+    >>> print(res)
+    --- Resources: ---
+    Total qubits: 3
+    Total gates : 279
+    Qubit breakdown:
+     clean qubits: 0, dirty qubits: 0, algorithmic qubits: 3
+    Gate breakdown:
+     {'Hadamard': 5, 'CNOT': 10, 'T': 264}
+
+    """
+    return _estimate_resources_dispatch(obj, gate_set, work_wires, tight_budget, config)
+
+
+@singledispatch
+def _estimate_resources_dispatch(
+    obj: ResourceOperator | Callable | Resources | list,
+    gate_set: set | None = None,
+    work_wires: int | dict = 0,
+    tight_budget: bool = False,
+    config: ResourceConfig | None = None,
+) -> Resources | Callable:
+    """Internal singledispatch function for resource estimation."""
+    raise TypeError(
+        f"Could not obtain resources for obj of type {type(obj)}. obj must be one of Resources, Callable or ResourceOperator"
+    )
+
+
+@_estimate_resources_dispatch.register
+def _resources_from_qfunc(
+    obj: Callable,
+    gate_set: set | None = None,
+    work_wires=0,
+    tight_budget=False,
+    config: ResourceConfig | None = None,
+) -> Callable:
+    """Get resources from a quantum function which queues operations"""
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        qm = WireResourceManager(work_wires, tight_budget)
+        # Get algorithm wires:
+        num_algo_qubits = 0
+        circuit_wires = []
+        for op in q.queue:
+            if isinstance(op, (ResourceOperator, Operation)):
+                if op.wires:
+                    circuit_wires.append(op.wires)
+                else:
+                    num_algo_qubits = max(num_algo_qubits, op.num_wires)
```

Comment:
Can't `num_wires` be `None`?

Reply:
I'm not actually too familiar with these functions -- still reviewing them myself, hence the WIP. If num_wires can be `None`, then we should just delete this else block.

Reasoning: If I had to guess, this is probably related to the wire management: I remember that we removed `num_wires` parameter from `ResourceOperator` that @Jaybsoni implemented for the `WireResourceManager`. Since that parameter is removed, I think we can safely just delete this else block -- likely relates to some functionality that Jay implemented in a prototype but didn't finalize.